[oneDNN] Design doc on oneDNN inplace pass (#2024)
PaddlePaddle/FluidDoc: doc/fluid/design/mkldnn/inplace/inplace.md

## Introduction

PaddlePaddle is implementing concept of in-place execution of some of operators.

> Idea of In-place execution is present on following picture:

Author: Good catch. thanks

![](images/inplace.svg)   

> Examplary graph presents three operators where one of them (type of elementwise_add) is to be performing in-place computation. In-place computation means that input variable(Tensor) is used for both input and output. This means that one of inputs will be overwritten with computational results. In presented picture in-place operator (elementwise_add) is 

Reviewer: Examplary -> Exemplary

Author: good catch!

> Exemplary graph presents three operators where one of them (type of elementwise_add) is to be performing in-place computation. In-place computation means that input variable(Tensor) is used for both input and output. This means that one of inputs will be overwritten with computational results. In presented picture in-place operator (elementwise_add) is 

Reviewer 2: ```suggestion
Exemplary graph presents three operators where one of them (type of elementwise_add) is to be performing in-place computation. In-place computation means that input variable (Tensor) is used for both input and output. This means that one of inputs will be overwritten with computational results. In presented picture in-place operator (elementwise_add) is 
```

> having two input nodes: *b* and *d*  and output *b*. So *b* is used for input and output and underneath it is represented by a one, shared Tensor. So this means that variable *b* is initially holding some input data and after the operator computation , input data is lost and replaced by computation's result.

Reviewer 2: ```suggestion
having two input nodes: *b* and *d* and output *b*. So *b* is used for input and output and underneath it is represented by a one, shared Tensor. So this means that variable *b* is initially holding some input data and after the operator computation, input data is lost and replaced by computation's result.
```

Currently assumption is that if operator can have in-place processing then all its kernel (including oneDNN) should be able to work properly in in-place mode. To match this functionality oneDNN integration was extended to support in-place execution for some of its operators:
- activations
- softmax
- elementwise_add
- gelu*
- sum**

> Adventages of in-place computation is:

Reviewer 2: ```suggestion
Adventages of in-place computation are:
```

> * lower memory usage.

Reviewer 2: ```suggestion
* lower memory usage,
```

* improved performance of operators

> To have in-place computation We need to analyze graph to search for where in-place execution could happen

Reviewer 2: ```suggestion
To have in-place computation, we need to analyze graph to search for where in-place execution could happen
```

and then make some of variables to be shared by input and output of in-place capable operator.

Hence there are two parts of in-place support:

> - in-place execution support within an operator
> - onednn inplace C-API pass

Reviewer 2: ```suggestion
- in-place execution support within an operator,
- onednn inplace C-API pass.
```

Author: not sure of this one as it is just a list , so I do not think commas and dot are needed

#### in-place execution support within an operator

> oneDNN primitive to have in-place execution needs to have same oneDNN memory object passed as input (src) and output(dst). In details we check if holded pointers to allocated buffers are the same for input and output

Reviewer 2: ```suggestion
For in-place execution, oneDNN primitive needs to have the same oneDNN memory object passed as input (src) and output (dst). More precisely, we check if pointers to allocated buffers are the same for input and output
```

> and this indicated if we use one oneDNN memory object or two. for example:

Reviewer 2: ```suggestion
and this indicates if we use one oneDNN memory object or two. For example:
```

`auto src_memory_p = handler.AcquireSrcMemory(x);`

`auto dst_memory_p = x->IsSharedBufferWith(*y) ? 
           src_memory_p : handler.AcquireDstMemory(y);`

#### oneDNN in-place pass

> As mentioned earlier idea of in-place pass is to locate operators with oneDNN kerenels that can perform in-place execution and then modify output node's variables to match input node's variable of the operator. 

Reviewer 2: ```suggestion
As mentioned earlier, idea of in-place pass is to locate operators with oneDNN kerenels that can perform in-place execution and then modify output node's variables to match input node's variable of the operator. 
```

##### Identifying operators with oneDNN kernels capable of in-place execution
This identification is a result of two checks:

> - Whether operator does have *inplaceInferer* structure

Author: There is a defition mentioned  below those two checks. But to rephrase it. each operator need to declare if it is capable of inplace exection or not. for example elementwise_add got two inputs X and Y and  output Z. Now inplaceinferer will be a strcutre (mapping) : X -> Z . This means that implementation can work in-place. And for elementwise_add Z (result) can replace X.

> - Whether operator does have *inplaceinferer* structure
> - Whether operator is on a list of oneDNN's in-place supported operators

Reviewer 2: ```suggestion
- Whether operator does have *inplaceInferer* structure,
- Whether operator is on a list of oneDNN's in-place supported operators.
```

Author: not sure of adding "," "." . will fix capitalization

*InplaceInferer* is a struct that declares a mapping (one of inputs to one of outputs) indicating that

> considerable operator can perform in-place execution and both vars (mentioned input and output in *InplaceInferer*) will

Reviewer 2: ```suggestion
considered operator can perform in-place execution and both vars (mentioned input and output in *InplaceInferer*) will
```

> share a tensor. This is not enough for oneDNN in-place C-API execution as oneDNN library may not provide in-place 

Author: Inplace execution is that  input and output are the same variable. So tensor is ofcourse the same as well. Using Holder() to check for inplace execution request is a way recommended by @sneaxiy and @luotao1 . When working on prototype I used ShareDataWith() , but you need to know to do it. E.g. information from graph (IR pass) is needed when to perform such operation. And graph 's passes operate on variables & nodes , tensors may not be allocated at that time.

> computation for all required (to have in-place execution) operators of PaddlePaddle and some of operators wouldb have to
> simulare in-place computation through the external buffer which would not bring any benefits, so there is no point enabling those in-place computations for C-API inference.

Reviewer 2: ```suggestion
computation for all required (to have in-place execution) operators of PaddlePaddle and some of operators would have to
simulate in-place computation through the external buffer which would not bring any benefits, so there is no point enabling those in-place computations for C-API inference.
```

##### Restrictions

> oneDNN in-place pass is taking adventage of graph pattern detector. So pattern consists of :
> Node (Var 1) -> Node(Op to be inplaced) -> Node(Var2) -> Node (next op after in-placed one) -> Node (Var3)
> Pattern is restricted so that in-placed to be op is of is oneDNN type. Due to fact that some operators have
> more than  one input and their output may be consumed by more than one operator it is expected that pattern
> maybe detected multiple times for the same operator e.g. once for one input, then for second intpu etc..

Reviewer 2: ```suggestion
oneDNN in-place pass is taking advantage of graph pattern detector. So pattern consists of:
Node (Var 1) -> Node (Op to be inplaced) -> Node (Var2) -> Node (next op after in-placed one) -> Node (Var3)
Pattern is restricted so that in-placed to be op is of oneDNN type. Due to fact that some operators have
more than one input and their output may be consumed by more than one operator it is expected that pattern
maybe detected multiple times for the same operator e.g. once for one input, then for second input etc...
```

Just having oneDNN operator capable of in-place is not enough to have in-place execution enabled, hence follwing rules
are checked by oneDNN in-place pass:
1. If input node to in-place operator is also an input to different operator, then in-place computation cannot be performed, as there is a risk that other operator consuming in-placed op operator will be executed after in-placed operator and therefore get invalid input data (overwritten by in-place computation).

> 2. If after in-placed operator there is another operator that is reusing in-place op's input var then in-place cannot happen unless next op can perform in-place computation. Next picture presents the idea.

Author: @luotao1 Actually I wanted input to be "b" . So this is entry situation like I can see in transformer model where some operators are reusing variables already used in previous operators. In that situation We need to be extra careful with applying in-place at graph , because we may *enforce* that some op like layer_norm is to work in-place, although it cannot.  So picture is showing situation before applying in-place. If I change input of layer norm to "a" then it would show situation that we would like to avoid. But intention of picture was to show entry situation.

> 1. If intput node to in-place operator is also an input to diffrent operator , then in-place computation cannot be performed , as there is a risk that other operator consuming in-placed op operator will be executed after in-placed operator and therfore get invalid input data (overwritten by in-place computation)
> 2. If after in-placed operator there is another operator that is reusing in-place op's input var then in-place cannot happen unless next op can perform in-place computation. Next picture presents the idea.

Reviewer 2: ```suggestion
1. If input node to in-place operator is also an input to different operator, then in-place computation cannot be performed, as there is a risk that other operator consuming in-placed op operator will be executed after in-placed operator and therefore get invalid input data (overwritten by in-place computation).
2. If after in-placed operator there is another operator that is reusing in-place op's input var then in-place cannot happen unless next op can perform in-place computation. Next picture presents the idea.
```

![](images/unwanted-inplace.svg)   

> In the picture we are seeing that in-place pass is considering to enable in-place execution for softmax oneDNN kernel. All is fine, but next operator after softmax is layer norm. Layer norm is already reusing input of softmax due to some earlier memory optimization pass being applied. If we make softmax op to perform in-place computation, then

Reviewer 3: > Layer norm is already reusing input of softmax due to some earlier memory optimization pass being applied.

Is this only an assumption? Not all layer norm has memory optimization pass being applied?

Author: I'm not targeting layer_norm , this is only an example. In-place pass is checking for any op that may reuse vars already used and then decide if in-place can be applied. So it is not only on layer norm

> So, in the picture we are seeing that in-place pass is considering to enable in-place execution for softmax oneDNN kernel. All is fine, but next operator after softmax is layer norm. Layer norm is already reusing input of softmax due to some earlier memory optimization pass being applied. If we make softmax op to perform in-place computation, then

Reviewer 2: ```suggestion
In the picture we are seeing that in-place pass is considering to enable in-place execution for softmax oneDNN kernel. All is fine, but next operator after softmax is layer norm. Layer norm is already reusing input of softmax due to some earlier memory optimization pass being applied. If we make softmax op to perform in-place computation, then
```

> it will also make layer norm to work in-place (b -> a). The thing is that layer norm cannot work in-place, so if we force it do so layer norm will produce invalid result.

Reviewer 3: > The thing is that layer norm cannot work in-place

Maybe this sentence makes confused, layer norm doesn't support in-place execution. You can remove this sentence.

Author: @luotao1 I'm not sure how to make it clearer. Why do you think quated sentence should be removed?

Reviewer 3: This sentence makes me feel that the reason is layer norm mkldnn kernel cannot support in-place execution, not some earlier memory optimization pass being applied.

Author: Situation here as described is that layer_norm is non-oneDNN , I have updated picture with this information. And lack of inplace execution for layer norm is detected by checking if InplaceInferer object exists (I added this info to PR). If InplaceInferer exists for operator after one we are to apply in-place then we can apply in-place modification , but we need to update also an output of next op .

##### In-place pass modification to graph when applied

When sub-graph is aligned with restrictions then in-place computation can be enabled. This is done by:

> 1. changing the name of output of in-place to be output node 

Reviewer: Didn't you mean 'changing the name of output of in-place to be input node' ?

Author: good catch!. thanks

> 1. changing the name of output node of in-place op to be match input node of in-place op.

Author: yes

> 1. changing the name of output node of in-place op to be match input node of in-place op.
> 2. renaming output var in output lists of node representing operator
> 3. Changing the name of input var in next op inputs list
> 4. If next Op is performing in-place computation then we need to updated next op's output as well not to break its
>    in-place computation.
> 5. if there are multiple operators after our in-place operator then we need to update all of them (their input vars). Idea is presented in the following picture:

Reviewer 2: ```suggestion
1. Changing the name of output node of in-place op to be match input node of in-place op.
2. Renaming output var in output lists of node representing operator.
3. Changing the name of input var in next op inputs list.
4. If next Op is performing in-place computation then we need to updated next op's output as well not to break its
   in-place computation.
5. if there are multiple operators after our in-place operator then we need to update all of them (their input vars). Idea is presented in the following picture:
```

![](images/multi-output-inplace.svg)   

We can see that there are two *top_k* operators after *elementwise_add* operator that is set to work in-placed. Each of *top_k* is having its own list of input vars, so we need to rename relevant input var to new name. As in-place pattern

> consists of: input node -> in-place op -> output node -> next op -> next op's output. For presented graph, there will be 8 patterns detected. Important thing is to remember original name of output, before it is renamed, so later we can

Reviewer 3: > 8 patterns

Could you describe what are 8 patterns?

Reviewer 3: Yes, it makes the PR more clearer.

Author: Ok, I updated PR with this info

> replace this orignal name in all of next op instanced.

Reviewer: This part could be more clear if You show both graphs, before and after inplace pass. Also then it would be better to use variables present on graph in explanation.

Author: ok, I have updated picture

> We can see that there are two *top_k* operators after *elementwise_add* operator that is set to work in-placed. Each of *top_k* is having its own list of input vars, so we need to rename relvant input var to new name. As in-place pattern
> consists of : input node -> in-place op -> output node -> next op -> next op's output. For presented there will be 8 patterns detected. Important thing is to remember original name of output , before it is renamed , so later we can
> replace this orignal name in all of next op instanced.

Reviewer 2: ```suggestion
We can see that there are two *top_k* operators after *elementwise_add* operator that is set to work in-placed. Each of *top_k* is having its own list of input vars, so we need to rename relevant input var to new name. As in-place pattern
consists of: input node -> in-place op -> output node -> next op -> next op's output. For presented graph, there will be 8 patterns detected. Important thing is to remember original name of output, before it is renamed, so later we can
replace this original name in all of next op instances.
```

> \* onednn gelu kernel is able to perform in-place execution , but currently gelu op does not support in-place support

Reviewer 2: ```suggestion
\* oneDNN gelu kernel is able to perform in-place execution, but currently gelu op does not support in-place.

```

> \*\* sum kernel is using oneDNN sum primitive that does not provide in-place exection , so in-place computation is done faked through external buffer. So it was not added into oneDNN inplace pass.

Reviewer 2: ```suggestion
\*\* sum kernel is using oneDNN sum primitive that does not provide in-place exection, so in-place computation is done faked through external buffer. So it was not added into oneDNN inplace pass.

```

##### Restrictions
oneDNN in-place pass is taking advantage of graph pattern detector. So pattern consists of:

> Node (Var 1) -> Node (oneDNN Op to be inplaced) -> Node (Var2) -> Node (next op - any typei, oneDNN/native CPU - after in-placed one) -> Node (Var3)

Reviewer 3: typei, is this a typo?

Author: yes, good catch. fixed
